patches/server/0814-Overridable-YAML-configuration.patch

```diff
@@ -0,0 +1,286 @@
+Date: Sun, 3 Oct 2021 08:03:16 +0200
+Subject: [PATCH] Overridable YAML configuration
+
+Co-authored-by: Bjarne Koll <lynxplay101@gmail.com>
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index f421e6a2e43e0a673dbb8a9a2b4331387e523e02..702477b94b28f0463fa51ee2aa999515457eaa68 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -18,6 +18,7 @@ import java.util.logging.Level;
+ import java.util.regex.Pattern;
+ 
+ import com.google.common.collect.Lists;
++import io.papermc.paper.configuration.OverridableYamlConfiguration;
+ import net.minecraft.server.MinecraftServer;
+ import org.bukkit.Bukkit;
+ import org.bukkit.ChatColor;
+@@ -55,7 +56,7 @@ public class PaperConfig {
+ 
+     public static void init(File configFile) {
+         CONFIG_FILE = configFile;
+-        config = new YamlConfiguration();
++        config = new OverridableYamlConfiguration("paper");
+         try {
+             config.load(CONFIG_FILE);
+         } catch (IOException ex) {
+diff --git a/src/main/java/io/papermc/paper/configuration/OverridableYamlConfiguration.java b/src/main/java/io/papermc/paper/configuration/OverridableYamlConfiguration.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..c0f29d6257ccd7ca2a6cd60e00a2ea62a3bb9062
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/configuration/OverridableYamlConfiguration.java
+@@ -0,0 +1,137 @@
++package io.papermc.paper.configuration;
++
++import com.google.common.base.Preconditions;
++import org.apache.commons.lang.Validate;
++import org.bukkit.Bukkit;
++import org.bukkit.configuration.InvalidConfigurationException;
++import org.bukkit.configuration.file.YamlConfiguration;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.io.File;
++import java.io.FileNotFoundException;
++import java.io.IOException;
++import java.util.logging.Level;
++
++/**
++ * An implementation of {@link YamlConfiguration} which allows overriding via {@link OverrideSource}.
++ */
++public class OverridableYamlConfiguration extends YamlConfiguration {
++    private final String group;
++    private final OverrideSource[] sources;
++
++    /**
++     * @param group   the group name of this configuration. This will be used as a prefix in sources.
++     *                The group name must only contain a-z, A-Z, and 0-9. It may contain underscore (_),
++     *                however, it must not end with underscore. Underscore and dots will be appended to
++     *                the group name in the respective sources.
++     * @param sources array of {@link OverrideSource} to fetch overridden values from.
++     */
++    public OverridableYamlConfiguration(@NotNull String group, @NotNull OverrideSource[] sources) {
++        Preconditions.checkArgument(group.matches("[a-zA-Z0-9_]*[a-zA-Z0-9]+"));
```

Comment:
can we precompile this regex in a static constant? its only called 3 times but still would look nicer.